cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll

```diff
@@ -0,0 +1,342 @@
+/**
+ * Provides implementation classes modeling C++ iterators, including
+ * `std::iterator`, `std::iterator_traits`, and types meeting the
+ * `LegacyIterator` named requirement. See `semmle.code.cpp.models.Models` for
+ * usage information.
+ */
+
+import cpp
+import semmle.code.cpp.models.interfaces.Taint
+import semmle.code.cpp.models.interfaces.DataFlow
+
+/**
+ * An instantiation of the `std::iterator_traits` template.
+ */
+class IteratorTraits extends Class {
+  IteratorTraits() {
+    this.hasQualifiedName("std", "iterator_traits") and
+    not this instanceof TemplateClass and
+    exists(TypedefType t |
+      this.getAMember() = t and
+      t.getName() = "iterator_category"
+    )
+  }
+
+  Type getIteratorType() { result = this.getTemplateArgument(0) }
+}
+
+/**
+ * A type which has the typedefs expected for an iterator.
+ */
+class IteratorByTypedefs extends Class {
+  IteratorByTypedefs() {
+    this.getAMember().(TypedefType).hasName("difference_type") and
+    this.getAMember().(TypedefType).hasName("value_type") and
+    this.getAMember().(TypedefType).hasName("pointer") and
+    this.getAMember().(TypedefType).hasName("reference") and
+    this.getAMember().(TypedefType).hasName("iterator_category") and
+    not this.hasQualifiedName("std", "iterator_traits")
+  }
+}
+
+/**
+ * The `std::iterator` class.
+ */
+class StdIterator extends Class {
+  StdIterator() { this.hasQualifiedName("std", "iterator") }
+}
+
+/**
+ * A type which can be used as an iterator
+ */
+class Iterator extends Type {
+  Iterator() {
+    this instanceof IteratorByTypedefs or
+    exists(IteratorTraits it | it.getIteratorType() = this) or
+    this instanceof StdIterator
+  }
+}
+
+/**
+ * A non-member prefix `operator*` function for an iterator type.
+ */
+class IteratorPointerDereferenceOperator extends Operator, TaintFunction {
+  IteratorPointerDereferenceOperator() {
+    this.hasName("operator*") and
+    this.getACallToThisFunction().getArgument(0).getFullyConverted().getUnderlyingType() instanceof
+      Iterator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A non-member `operator++` or `operator--` function for an iterator type.
+ */
+class IteratorCrementOperator extends Operator, DataFlowFunction {
+  IteratorCrementOperator() {
+    (
+      this.hasName("operator++") or
+      this.hasName("operator--")
```

Comment:
I believe they are both called `operator++` (or `operator--`) but have slightly different parameters to distinguish them (a dummy `int` parameter is added to one of them).  It's an odd little language quirk.  We test for the taint flow in these two test cases:
```
sink(*(++i3)); // tainted
```
and
```
sink(*(--i4)); // tainted
```
So I wouldn't mind seeing another two test cases added.

Reply:
The above related to code added by an older PR, in the string.cpp taint test; assuming @rdmarsh2 hasn't addressed this I don't mind making a separate PR to add the other two cases.

It would be good to have answers to the other questions that relate to the older PR - even though it's probably not appropriate to fix things here I think they are still good questions.

Reply:
I've added the two test missing test cases I was talking about in https://github.com/github/codeql/pull/4394.